File: src/main/java/com/intgrah/algorithms/heap/BinomialHeap.java
package com.intgrah.algorithms.heap;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class BinomialHeap<K> extends DecreasableHeap<K> {

    private final List<Node> roots = new ArrayList<>();
    private Node min = null;

    public BinomialHeap(Comparator<K> ord) { super(ord); }

    @Override
    public Decreasable pushRef(K k) {
        Node n = new Node(k);
        addTree(n);
        if (min == null || ord.compare(n.key, min.key) < 0)
            min = n;
        return n;
    }

    private void addTree(Node n) {
        int deg = n.degree();
        while (deg < roots.size() && roots.get(deg) != null)
            n = link(n, roots.set(deg++, null));
        if (deg == roots.size())
            roots.add(n);
        else
            roots.set(deg, n);
    }

    private Node link(Node m, Node n) {
        if (ord.compare(m.key, n.key) < 0) {
            m.children.add(n);
            n.parent = m;
            return m;
        } else {
            n.children.add(m);
            m.parent = n;
            return n;
        }
    }

    @Override
    public K getMin() {
        assert !isEmpty();
        return min.key;
    }

    @Override
    public void deleteMin() {
        assert !isEmpty();
        roots.set(min.degree(), null);
        for (Node n : min.children) {
            n.parent = null;
            addTree(n);
        }
        min = null;
        for (Node n : roots)
            if (n != null && (min == null || ord.compare(n.key, min.key) < 0))
                min = n;
    }

    @Override
    public boolean isEmpty() { return min == null; }

    @Override
    public void clear() {
        min = null;
        roots.clear();
    }

    private class Node extends Decreasable {

        private List<Node> children = new ArrayList<>();
        private Node parent = null;

        private Node(K k) { super(k); }

        @Override
        public void decreaseKey(K k) {
            assert ord.compare(k, key) <= 0;
            key = k;
            if (ord.compare(k, min.key) < 0)
                min = this;
            while (parent != null && ord.compare(k, parent.key) < 0) {
                Node p = parent;
                Node g = p.parent;
                int nDeg = degree();
                int pDeg = p.degree();
                p.children.set(nDeg, p);
                if (g == null)
                    roots.set(pDeg, this);
                else
                    g.children.set(pDeg, this);
                parent = g;
                p.parent = this;
                List<Node> pChildren = p.children;
                for (var child : children)
                    child.parent = p;
                for (Node child : pChildren)
                    child.parent = this;
                p.children = children;
                children = pChildren;
            }
        }

        private int degree() { return children.size(); }

    }

}

File: heap/binomial_heap.py
from __future__ import annotations


class BinomialNode[K]:
    def __init__(self, key: K) -> None:
        self.key = key
        self.parent: BinomialNode[K] | None = None
        self.children = list[BinomialNode[K]]()

    def merge(self, n: BinomialNode[K]) -> BinomialNode[K]:
        if self.key < n.key:
            n.parent = self
            self.children.append(n)
            return self
        else:
            self.parent = n
            n.children.append(self)
            return n


class BinomialHeap[K]:
    def __init__(self) -> None:
        self.min: BinomialNode[K] | None = None
        self.roots: list[BinomialNode[K]] = []

    def add(self, n: BinomialNode[K]) -> None:
        while True:
            if (deg := len(n.children)) >= len(self.roots):
                for _ in range(deg - len(self.roots)):
                    self.roots.append(None)
                self.roots.append(n)
                return
            n2 = self.roots[deg]
            if n2 is None:
                self.roots[deg] = n
                return
            n = n.merge(n2)
            self.roots[deg] = None

    def push(self, key: K) -> BinomialNode[K]:
        self.add(n := BinomialNode(key))
        if self.min is None or n.key < self.min.key:
            self.min = n
        return n

    def pop_min(self) -> K:
        if (min := self.min) is None:
            raise IndexError
        self.roots[len(min.children)] = None
        for n in min.children:
            n.parent = None
            self.add(n)
        self.min = None
        for n in self.roots:
            if n is not None and (self.min is None or n.key < self.min.key):
                self.min = n
        return min.key

    def decrease_key(self, n: BinomialNode[K], k: K) -> None:
        assert k <= n.key
        n.key = k
        if k < self.min.key:
            self.min = n
        while (p := n.parent) is not None and k < p.key: 
            # Swap n and p
            t_deg = len(n.children)
            p_deg = len(p.children)
            p.children[t_deg] = p
            if p.parent is None:
                self.roots[p_deg] = n
            else:
                p.parent.children[p_deg] = n
            n.parent, p.parent = p.parent, n
            n.children, p.children = p.children, n.children

    def __bool__(self) -> bool:
        return self.min is not None
